Support More Interpolations on SplinePlug

danieldresser-ie opened the pull request:
This is currently very preliminary - it definitely doesn't work, and I haven't even started on adding support for MonotoneCubic, it's just the preliminary refactor.

I just wanted to check that this looks like the right direction before I take it further.

The only significant difference from what we discussed via email is that instead of "inferValue" which constructs a list of points from a Cortex spline, I've currently just got trimEndPoints, which constructs a list of points from a list of points with duplicated end points.  None of the places where I needed inferValue were actually starting from a Cortex spline, so it felt a bit silly to go: "OSL basis string name -> Cortex spline basis matrix -> Interpolation enum" instead of just  "OSL basis string name -> Interpolation enum", and then process the points.  Happy to do it the other way if you prefer.

Also, not sure what to name getValueAsCortexSpline().

johnhaddon commented on src/Gaffer/SplinePlug.cpp:
```diff
@@ -106,45 +99,48 @@ bool SplinePlug<T>::acceptsChild( const GraphComponent *potentialChild ) const
+	Ptr result = new SplinePlug<T>( name, direction, m_defaultPoints, m_defaultInterpolation, getFlags() );
+	typename T::PointContainer points;
+	typename SplinePlug<T>::Interpolation interpolation;
+	getValue( points, interpolation );
```
Hmm, `createCounterpart` isn't expected to transfer the current value, just to create a fresh plug with the same defaults. I see you're just mirroring what was there already, but I don't recall a reason why SplinePlug should have been different. Could you try pulling out that `setValue()` call, perhaps in a separate, preceding commit?

johnhaddon commented on src/GafferOSL/OSLShader.cpp:
```diff
@@ -659,30 +659,38 @@ template <typename PlugType>
 
 	const std::string &basis = basisParameter->sdefault.front().string();
+	typename PlugType::Interpolation defaultInterpolation = PlugType::CatmullRom;
+	// TODO - should we keep low level support of Bezier even if we don't want to expose it in the UI?
```
Maybe remove it? It can't have been used successfully in the past, so I don't see any harm in removing it...

johnhaddon commented on include/Gaffer/SplinePlug.h:
```diff
@@ -72,10 +72,20 @@ class SplinePlug : public ValuePlug
 			const std::string &name = defaultName<SplinePlug>(),
 			Direction direction=In,
-			const T &defaultValue = T(),
+			const typename T::PointContainer &defaultPoints = T::PointContainer(),
```
How are you feeling about using two separate variables for the default and value? I'm wondering if we shouldn't wrap them up in a simple little `SplinePlug::ValueType` struct. This would also fix the problem that the current `ValueType` typedef is inaccurate - advertising that `Spline` is the value type.

danieldresser-ie replied:
If we were to do this SplinePlug::ValueType struct, would the SplinePlug still be templated on an IECore::Spline?

It kind of seems like it could just be templated on the Color3f or float, and then spline() would return IECore::Spline<float, T>.  If we're overhauling everything, it seems a bit weird to keep IECore::Spline as the thing it's templated on, if it's just one member function that does anything with IECore::Spline.

johnhaddon commented on include/Gaffer/SplinePlug.h:
```diff
@@ -85,25 +95,23 @@ class SplinePlug : public ValuePlug
+		/// Matching to setValue
+		void getValue( typename T::PointContainer &points, Interpolation &interpolation ) const;
+
+		T getValueAsCortexSpline() const;
```
I think I would either just call this `spline()`, or possibly add it as a method on the `ValueType` I proposed, so you could do `plug->getValue().spline()`.

johnhaddon commented on src/GafferScene/LightTweaks.cpp:
```diff
@@ -256,6 +256,28 @@ void numericTweak( Data *data, PlugType *plug, LightTweaks::TweakPlug::Mode mode
 }
 
+template<typename PlugType>
+void splineTweak( Data *data, PlugType *plug, LightTweaks::TweakPlug::Mode mode )
```
How do you feel about this? It seems a bit disappointing that we need a special case here - I guess it's highlighting the fact that we are deviating a bit from the other ValuePlugs in our SplinePlug design. The only way I can see that might work is if our new `SplinePlug::ValueType` struct provided an implicit conversion to `Spline` - what do you think to that?

johnhaddon commented on include/Gaffer/SplinePlug.h:
```diff
@@ -124,14 +132,17 @@ class SplinePlug : public ValuePlug
-		const IntPlug *endPointMultiplicityPlug() const;
+		// If you are starting with a curve representation that needs duplicated end point values, and you're
+		// converting it into this plug representation, you should trim off the duplicated end point values first
+		static bool trimEndPoints( typename T::PointContainer &points, Interpolation interpolation );
```
Do we need this anywhere apart from the shader loaders? If not, I'd be tempted to squirrel it away as a protected method on the Shader base class.

danieldresser-ie commented on include/Gaffer/SplinePlug.h:
```diff
@@ -72,10 +72,20 @@ class SplinePlug : public ValuePlug
 
+		// TODO - this is where John had suggested this go in his initial email, but does it
+		// really make for this enum to be part of each separate templated version of SplinePlug?
+		enum Interpolation
```
John, what do you think of this Interpolation going here?  Do you think it's an advantage to have many different versions of the Interpolation enum floating around?  I guess it makes the namespacing easier.

danieldresser-ie commented:
OK, I've now taken a stab at reworking everything to have a struct that holds interpolation and points.  Unfortunately, I didn't catch you this morning with my question about whether I SplinePlug should now be templated just on the Y type instead of IECore::Spline.  Don thought it seemed like a good idea, so I went ahead and tried that - hopefully this has moved things somewhat in the right direction.

The struct is currently named SimpleSpline, this was a bit of a placeholder name while I dealt with things using the old meaning of ValueType, but I guess I can rename it to ValueType now.

The main part I'm not sure about is that I now need to create Python bindings for SplinefColor3fPlug::ValueType - do I just name these SplinefColor3fPlug_ValueType or something?  I got to the point of needing to bind a constructor and serializer for my current SplinefColor3fPlug_SimpleSpline, and then started wondering whether this was overcomplicating things, and the binding for getValue should translate to a tuple.  But we probably should be able to call spline() on a ValueType from Python, so I guess in that case spline() would have to become a static function on SplinePlug.

One thing this didn't help at all with is the special case in splineTweak - basicTweak relies on being able to instantiate a  TypedData<typename PlugType::ValueType> - that's not going to be what we want unless SplinePlug lies and says it's ValueType is IECore::Spline.

Oh, and a question - the widget for the interpolation needs to be a dropdown on RampPlugValueWidget, but we can't currently make PresetsPlugValueWidgets without using metadata?  The simplest here would be to just use an EnumPlugValueWidget, but are those supposed to be deprecated?

johnhaddon commented:
> Unfortunately, I didn't catch you this morning with my question about whether I SplinePlug should now be templated just on the Y type instead of IECore::Spline. Don thought it seemed like a good idea, so I went ahead and tried that - hopefully this has moved things somewhat in the right direction.

Unfortunately I feel differently here, and think it should stay as `SplinePlug<SplineType>`. There might only be one method to get the spline, but getting a spline is the exact reason I would be using one of these plugs, so to me it makes sense to put the spline type front and centre. It also avoids ambiguity about the type for the X axis (which you assumed to be float in your example), and makes the use of SplineType::PointsContainer a bit more obvious.

> The struct is currently named SimpleSpline, this was a bit of a placeholder name while I dealt with things using the old meaning of ValueType, but I guess I can rename it to ValueType now.

One other thought I had is that if what you're calling SimpleSpline is the "user facing, reasonably easily editable spline type", then at some point it might belong in Cortex as a companion to the "mathsy, suitable for evaluation" Spline type, or at least it might be useful outside of SplinePlug. In that case it might be natural to template SplinePlug on the SimpleSpline type rather than the Spline, which gets us back some consistency with other plug types, because then `SplinePlug<T>::getValue()` would once again be returning a `T`.

> The main part I'm not sure about is that I now need to create Python bindings for SplinefColor3fPlug::ValueType - do I just name these SplinefColor3fPlug_ValueType or something? I got to the point of needing to bind a constructor and serializer for my current SplinefColor3fPlug_SimpleSpline, and then started wondering whether this was overcomplicating things, and the binding for getValue should translate to a tuple. But we probably should be able to call spline() on a ValueType from Python, so I guess in that case spline() would have to become a static function on SplinePlug.

If they stay as a nested class and not as a standalone class, then we should bind them as nested classes of the python class. This is pretty straightforward using a `boost::python::scope` variable - the [Shuffle bindings](https://github.com/GafferHQ/gaffer/blob/master/src/GafferImageModule/ImageProcessorBinding.cpp#L150) are a reasonable example. But maybe they should be a standalone class?

> One thing this didn't help at all with is the special case in splineTweak - basicTweak relies on being able to instantiate a TypedData - that's not going to be what we want unless SplinePlug lies and says it's ValueType is IECore::Spline.

Yeah, the only way I see us improving that situation is to _really_ make SimpleSpline into a standalone class, have SimpleSplineData too, and then do a final conversion in the renderer backends. This doesn't sound like a terrible idea conceptually, but it seems like a bit of a rabbit hole with no immediate benefit. So I guess we stick with splineTweak for now?

> Oh, and a question - the widget for the interpolation needs to be a dropdown on RampPlugValueWidget, but we can't currently make PresetsPlugValueWidgets without using metadata?

Correct. The only way we could really register metadata to it would be if it had it's own InterpolationPlug type, or alternatively if we made it possible to register metadata to named descendants of plug types in the same way we can for named descendants of node types.

> The simplest here would be to just use an EnumPlugValueWidget, but are those supposed to be deprecated?

I haven't commented it as being deprecated, but I have been steadily removing their use in the hope of getting rid of it one day - we're down to one now in Gaffer, that'll be removed when we drop RSL shaders. Every place EnumPlugValueWidgets have been replaced by metadata I think we're better off, because metadata interacts nicely with everything else (documentation, the ui editor etc).

The other problem with not being able to register metadata to the child plug is that if it were to be promoted on its own, it wouldn't have any way of taking it's widget with it. Like the current basis widget, it's only going to work at all because the RampPlugValueWidget will explicitly construct a widget of the right type rather than let metadata specify it.

I guess we have :

1. Add a mechanism for registering metadata to child plugs. Possibly useful elsewhere, but a bunch of work we shouldn't lump in to this PR.
2. Make the interpolation plug have its own distinct type. The logical end of this approach is an explosion of new plug types for everywhere we use an enum.
3. Use EnumPlugValueWidget, and hope to remove it in future somehow after doing 1.
4. Use PresetsPlugValueWidget, and cheekily register the presets just prior to construction in RampPlugValueWidget. This is definitely a bit ugly but at least doesn't stand in the way of the other refactoring.

danieldresser-ie commented:
OK, sounds like the direction you want is to make SimpleSpline a standalone class.

What should it be named, and where should it live?

I guess you do want to be able to template the X type on it as well?  Don felt that hardcoding to float was clearer, and neither of us could really think of any use for a double precision spline, and he said he had successfully been ripping out unused double precision paths in some of his Cortex work, so I thought I could get away with it.

As for the presets thing, I don't feel strongly about it, as long as it works somehow - it sounds like you're leaning towards 4?

johnhaddon commented:
> OK, sounds like the direction you want is to make SimpleSpline a standalone class.

Does this seem reasonable to you? I'm figuring this out as we go, so if it doesn't, then do say. The thing I like most about it is that it gets us back to `T SplinePlug<T>::getValue()`, so SplinePlug will continue to match the pattern established for all other plug types.

> What should it be named, and where should it live?

Good question. Any thoughts? For now I think it could probably live happily enough above the SplinePlug in the same header? Not sure what it's called though. What do we think it is? A spline proxy? A spline definition? A conveniently editable form of a spline? A gaffer spline?

> guess you do want to be able to template the X type on it as well?

Since it's purpose is to define a Spline, I think it's probably clearer and easier to template it on the Spline type.

> Don felt that hardcoding to float was clearer, and neither of us could really think of any use for a double precision spline, and he said he had successfully been ripping out unused double precision paths in some of his Cortex work, so I thought I could get away with it.

I agree we probably don't need to use a double spline - we've managed to avoid DoublePlug thus far. But I think it's clearer and easier to template on the spline type (the thing we are making the class to represent), and it keeps the door open for the future. Note that unlike the double paths Don has been ripping out, we won't be paying anything here, because we're not going to instantiate the template for doubles.

> As for the presets thing, I don't feel strongly about it, as long as it works somehow - it sounds like you're leaning towards 4?

I don't feel too strongly either - 3 or 4 will do.

danieldresser-ie commented:
> Not sure what it's called though. 

Hmm, nothing I really like, but SplineDefinition feels like the best option so far.

> Since it's purpose is to define a Spline, I think it's probably clearer and easier to template it on the Spline type.

Agreed.  My somewhat unclear question is whether SplineDefinition needs to also be doubly templated on X and Y.  It sounds like you would be OK if SplineDefinition always used float for the X axis?

danieldresser-ie commented:
Err, I was in a hurry to reply and just realized the second half of my previous reply is nonsensical.  You're talking about templating SplineDefinition on Spline.  I guess I can do that.  Contemplating ...

johnhaddon commented:
> Hmm, nothing I really like, but SplineDefinition feels like the best option so far.

OK, let's run with that.

> My somewhat unclear question is whether SplineDefinition needs to also be doubly templated on X and Y. It sounds like you would be OK if SplineDefinition always used float for the X axis?

This is what I was thinking, but probably not communicating very well...

```
/// Templated on the type of the spline being defined
template<typename T>
class SplineDefinition
{
     typedef T SplineType;
     ...
     SplineType spline();
}

typedef SplineDefinition<Splineff> SplineDefinitionff;
typedef SplineDefinition<SplinefColor3f> SplineDefinitionfColor3f;
```

I think that makes sense right?

Then for SplinePlug itself we either have this :

```
template<typename T>
class SplinePlug
{
     typedef T ValueType;
     ...
     void setValue( const T &value );
     T getValue() const;
     ...    
}

typedef SplinePlug<SplineDefinitionfColor3f> SplinefColor3fPlug;
```

or this :

```
template<typename T>
class SplinePlug
{
     typedef SplineDefinition<T> ValueType;
     ...
     void setValue( const ValueType &value );
     ValueType getValue() const;
     ...    
}

typedef SplinePlug<SplinefColor3f> SplinefColor3fPlug;
```

Of those two I was originally thinking the former, but I seem to be a bit tied to the latter, because that's what I've been used to so far...

danieldresser-ie commented:
I was definitely thinking the former.

When it comes to how SplineDefinition is templated, both Don and I find it pretty weird to be templated on IECore::Spline.  SplineDefinition is simpler than Spline, and it really feels way simpler to me to just template it on Color3f or float.

This doesn't seem to have too much overall impact on things though, so I can go ahead and do it that way if it seems better to you.

johnhaddon commented:
> When it comes to how SplineDefinition is templated, both Don and I find it pretty weird to be templated on IECore::Spline.

What could be more natural than a spline definition being templated on the type of spline it will define?

> This doesn't seem to have too much overall impact on things though, so I can go ahead and do it that way if it seems better to you.

It does seem better to me, yes.

In case you think we're going down a rabbit hole here though, I'll throw out a simpler option. We originally got to the SplineDefinition because the `getValue()/setValue()` methods in the original prototype felt awkward, and had different python bindings than the C++ (tuple return type). If we want to avoid SplineDefinition entirely then we could drop `setValue()/getValue()` and instead have a `setPoints/getPoints` pair and either a `setInterpolation()/getInterpolation()` pair or use the interpolation plug directly.

danieldresser-ie commented:
OK, everything seems to be working now, including the compatibility config, and tests appear to be passing.

The only thing I haven't done yet is switch SplineDefinition to be templated on IECore::Spline ( I still feel like this makes things worse, but it should hopefully be straightforward to do. ).

johnhaddon commented on python/GafferUI/RampPlugValueWidget.py:
```diff
@@ -52,8 +52,16 @@ def __init__( self, plug, **kw ) :
+				# TODO: Since we don't have a good way to register metadata on child plugs, we just write the
+				# metadata on this child plug right before constructing a widget for it.  There should probably
+				# be some way to do this genericly during initialization
+				Gaffer.Metadata.registerValue( plug['interpolation'],
```
Can we pass `persistent = False` here so we don't pollute the serialisations with the metadata? I think that should make things easier to tidy up in the future if we get the registration mechanism we really need.

johnhaddon commented on src/GafferBindings/SplinePlugBinding.cpp:
```diff
@@ -54,6 +55,89 @@ using namespace Gaffer;
 {
 
+template<typename T>
+std::string splineDefinition_repr( T &x )
```
Camel case please

johnhaddon commented on src/GafferBindings/SplinePlugBinding.cpp:
```diff
@@ -54,6 +55,89 @@ using namespace Gaffer;
+		.def( "__repr__", &splineDefinition_repr<T> )
+		.def( "points", &splineDefinition_points<T>, "Read only access to the control points as a tuple of tuples of ( x, y ) pairs." )
+		.def_readwrite("interpolation", &T::interpolation)
+        .def( self==self )
```
Tabs please

johnhaddon commented on src/GafferBindings/SplinePlugBinding.cpp:
```diff
@@ -54,6 +55,89 @@ using namespace Gaffer;
+}
+
+
+#define REPR_SPLINEDEFINITION_SPECIALISATION( TYPE )                                                    \
```
Looks like this approach is taken from the Cortex spline bindings? If you're up for it, I think we can do a bit better than resorting to macros to replace a default implementation that silently fails :

```
template<typename T>
std::string splineDefinitionRepr( object x )
{
    string stream s;
    /// Query name from python object itself, so we don't need the macro.
    s << "Gaffer." << extract<std::string>( object.attr( "__name__" ) ) << "( ";
    /// Use extract to get the spline ready for extracting the points etc.
    const T splineDefinition = extract<T>( x );
    ...
}
```

danieldresser-ie commented:
OK, I think that covers the feedback, and the tests are now passing, aside from a deadlock in something related to the ImageCatalog, which I'm pretty confident isn't related.  There appear to be some test failures across the board right now related to Catalog/Display?

This isn't very useful without actually implementing the MonotoneCubic type - not sure whether you want to try and actually merge this first - otherwise if this looks good for a start, I'll upload some MonotoneCubic stuff tomorrow.

johnhaddon commented:
Oh, wait, one other thing I forgot to mention. In the compatibility config I don't see anything dealing with calls to `splinePlug["basis"]["matrix"].setValue()` - I think we'll need something to translate that into an interpolation now won't we? Otherwise any files with non-default interpolation will be broken?

danieldresser-ie commented:
OK, it looks like this is actually working now.

On further investigation, it looks like the interpolation used by the ramp in Htoa is pretty reasonable in practice for editing ramps, so we can probably just aim for full compatibility with that.

The only weird thing is that CatmullRom and MonotoneCubic could be expected to be the same except for the monotone constraint, but in practice they are different in several ways, because the underlying parameterization of the X values is linear for MonotoneCubic.  Doing CatmullRom the same way might actually be a more useful curve type for color ramps, but we would then need to store it as a bezier curve as well.  ( Ideally for this interpolation, we would be storing the X values with a different interpolation type, rather than using a degenerate bezier curve that is costly to invert ).

But I should probably ignore that quibble, and consider this final functionality for the moment.

There will probably be some code rearrangement needed.  I added a test to OSLShaderTest that actually evaluates a Spline parameter and makes sure it matches Cortex's spline evaluation.  This test will currently fail in master until I finish up my Cortex PR.

The last commit where I add a function to account for differences between our spline type and OSL, I had no idea where you would want this to live, so I haven't made any effort.  Pretty sure you won't want it in OSLShader.h, but this is enough to confirm the functionality is working.

I've also added commits to address the last round of comments, and haven't squashed them yet.

danieldresser-ie commented:
Squashed last round of fixes to be ready to merge.
